In [20]:
import pandas as pd


In [21]:
df = pd.read_csv('data/train.csv')

In [22]:
df['Outcome linked to object of search'] = df['Outcome linked to object of search'].fillna(False)

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 856610 entries, 0 to 856609
Data columns (total 16 columns):
 #   Column                                    Non-Null Count   Dtype  
---  ------                                    --------------   -----  
 0   observation_id                            856610 non-null  object 
 1   Type                                      856610 non-null  object 
 2   Date                                      856610 non-null  object 
 3   Part of a policing operation              646620 non-null  object 
 4   Latitude                                  700308 non-null  float64
 5   Longitude                                 700308 non-null  float64
 6   Gender                                    856610 non-null  object 
 7   Age range                                 856610 non-null  object 
 8   Self-defined ethnicity                    841427 non-null  object 
 9   Officer-defined ethnicity                 856610 non-null  object 
 10  Legislation         

In [24]:
df

observation_id                       Type  \
0       2e4d0094-c30b-471b-a211-72a9790feca2              Person search   
1       4779fbe8-6e05-4534-85fd-db32952ee309              Person search   
2       cb5c685d-acac-42e2-914d-75e6ff73b0a8              Person search   
3       f486e116-5b1e-45db-9931-a7f070c5c478              Person search   
4       78f4020e-12cc-4889-bf1a-2f2c29b2f662              Person search   
...                                      ...                        ...   
856605  ee337b9a-12ad-45fd-8c60-49091a0f4ab8  Person and Vehicle search   
856606  5973a004-e579-4dd2-bc26-71ab5717f87a  Person and Vehicle search   
856607  ad053a34-364e-4d24-8f5c-9734ab5fdbe0  Person and Vehicle search   
856608  8736e5ec-7ca2-420b-ad56-4dd88d27fe6e              Person search   
856609  a697ebdf-0ff2-4d6e-8c0e-1e6285858301              Person search   

                             Date Part of a policing operation   Latitude  \
0       2020-12-01T01:10:00+00:00                          NaN  50.798824   
1       2020-12-01T02:00:00+00:00                          NaN  50.785099   
2       2020-12-01T09:15:00+00:00                          NaN  50.952006   
3       2020-12-01T10:20:00+00:00                          NaN  50.806383   
4       2020-12-01T10:24:00+00:00                          NaN  50.806670   
...                           ...                          ...        ...   
856605  2020-04-30T15:10:00+00:00                          NaN  54.965502   
856606  2020-04-30T15:10:00+00:00                          NaN  54.965502   
856607  2020-04-30T17:00:00+00:00                          NaN  54.966266   
856608  2020-04-30T17:35:00+00:00                          NaN  54.971596   
856609  2020-04-30T17:36:00+00:00                          NaN  54.972352   

        Longitude Gender Age range  \
0       -1.089471   Male     25-34   
1       -1.091540   Male   over 34   
2       -1.403341   Male   over 34   
3       -1.079844   Male     10-17   
4       -1.081982   Male     10-17   
...           ...    ...       ...   
856605  -1.604609   Male     18-24   
856606  -1.604609   Male     25-34   
856607  -1.453704   Male     18-24   
856608  -1.636589   Male     25-34   
856609  -1.637175   Male     18-24   

                                   Self-defined ethnicity  \
0                         Other ethnic group - Not stated   
1                      White - Any other White background   
2       White - English/Welsh/Scottish/Northern Irish/...   
3                         Other ethnic group - Not stated   
4                         Other ethnic group - Not stated   
...                                                   ...   
856605                 White - Any other White background   
856606                 White - Any other White background   
856607  White - English/Welsh/Scottish/Northern Irish/...   
856608  White - English/Welsh/Scottish/Northern Irish/...   
856609  White - English/Welsh/Scottish/Northern Irish/...   

       Officer-defined ethnicity  \
0                          White   
1                          Other   
2                          White   
3                          White   
4                          Asian   
...                          ...   
856605                     White   
856606                     Other   
856607                     White   
856608                     White   
856609                     White   

                                              Legislation  \
0       Police and Criminal Evidence Act 1984 (section 1)   
1                   Misuse of Drugs Act 1971 (section 23)   
2                   Misuse of Drugs Act 1971 (section 23)   
3       Police and Criminal Evidence Act 1984 (section 1)   
4       Police and Criminal Evidence Act 1984 (section 1)   
...                                                   ...   
856605              Misuse of Drugs Act 1971 (section 23)   
856606              Misuse of Drugs Act 1971 (section 23)   
856607              Misuse of Dr

In [25]:
df.columns

Index(['observation_id', 'Type', 'Date', 'Part of a policing operation',
       'Latitude', 'Longitude', 'Gender', 'Age range',
       'Self-defined ethnicity', 'Officer-defined ethnicity', 'Legislation',
       'Object of search', 'Outcome', 'Outcome linked to object of search',
       'Removal of more than just outer clothing', 'station'],
      dtype='object')

In [26]:
df_categorical_columns = (df.drop(columns=
        [
            'observation_id', 
            'Type', 
            'Date', 
            'Part of a policing operation', 
            'Latitude', 
            'Longitude',
        ])
)

In [27]:
df_categorical_columns

Gender Age range                             Self-defined ethnicity  \
0        Male     25-34                    Other ethnic group - Not stated   
1        Male   over 34                 White - Any other White background   
2        Male   over 34  White - English/Welsh/Scottish/Northern Irish/...   
3        Male     10-17                    Other ethnic group - Not stated   
4        Male     10-17                    Other ethnic group - Not stated   
...       ...       ...                                                ...   
856605   Male     18-24                 White - Any other White background   
856606   Male     25-34                 White - Any other White background   
856607   Male     18-24  White - English/Welsh/Scottish/Northern Irish/...   
856608   Male     25-34  White - English/Welsh/Scottish/Northern Irish/...   
856609   Male     18-24  White - English/Welsh/Scottish/Northern Irish/...   

       Officer-defined ethnicity  \
0                          White   
1                          Other   
2                          White   
3                          White   
4                          Asian   
...                          ...   
856605                     White   
856606                     Other   
856607                     White   
856608                     White   
856609                     White   

                                              Legislation  \
0       Police and Criminal Evidence Act 1984 (section 1)   
1                   Misuse of Drugs Act 1971 (section 23)   
2                   Misuse of Drugs Act 1971 (section 23)   
3       Police and Criminal Evidence Act 1984 (section 1)   
4       Police and Criminal Evidence Act 1984 (section 1)   
...                                                   ...   
856605              Misuse of Drugs Act 1971 (section 23)   
856606              Misuse of Drugs Act 1971 (section 23)   
856607              Misuse of Drugs Act 1971 (section 23)   
856608              Misuse of Drugs Act 1971 (section 23)   
856609              Misuse of Drugs Act 1971 (section 23)   

                Object of search                       Outcome  \
0       Article for use in theft          Community resolution   
1               Controlled drugs  A no further action disposal   
2               Controlled drugs  A no further action disposal   
3                   Stolen goods  A no further action disposal   
4              Offensive weapons  A no further action disposal   
...                          ...                           ...   
856605          Controlled drugs  A no further action disposal   
856606          Controlled drugs  A no further action disposal   
856607          Controlled drugs      Khat or Cannabis warning   
856608          Controlled drugs                        Arrest   
856609          Controlled drugs                        Arrest   

        Outcome linked to object of search  \
0                                    False   
1                                    False   
2                                    False   
3                                    False   
4                                    False   
...                                    ...   
856605                                True   
856606                                True   
856607                                True   
856608                                True   
856609                                True   

       Removal of more than just outer clothing      station  
0                                         False    hampshire  
1                                         False    hampshire  
2                                          True    hampshire  
3                                         False    hampshire  
4                                         False    hampshire  
...                                         ...          ...  
856605                                    False  northumbria  
856606                                     True  north

In [28]:
df_categorical_columns.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 856610 entries, 0 to 856609
Data columns (total 10 columns):
 #   Column                                    Non-Null Count   Dtype 
---  ------                                    --------------   ----- 
 0   Gender                                    856610 non-null  object
 1   Age range                                 856610 non-null  object
 2   Self-defined ethnicity                    841427 non-null  object
 3   Officer-defined ethnicity                 856610 non-null  object
 4   Legislation                               816863 non-null  object
 5   Object of search                          856610 non-null  object
 6   Outcome                                   856610 non-null  object
 7   Outcome linked to object of search        856610 non-null  bool  
 8   Removal of more than just outer clothing  340097 non-null  object
 9   station                                   856610 non-null  object
dtypes: bool(1), object(9)
memory usa

In [34]:
df_categorical_columns[['Outcome linked to object of search', 'Outcome', ]].value_counts().sort_index()

Outcome linked to object of search  Outcome                        
False                               A no further action disposal       582321
                                    Arrest                              80876
                                    Caution (simple or conditional)       994
                                    Community resolution                48140
                                    Khat or Cannabis warning             2417
                                    Penalty Notice for Disorder         16084
                                    Summons / charged by post           10218
True                                A no further action disposal        45153
                                    Arrest                              30872
                                    Caution (simple or conditional)      2265
                                    Community resolution                20140
                                    Khat or Cannabis warning             9

In [29]:
(df_categorical_columns.Gender.value_counts()/len(df_categorical_columns.Gender))

Male      0.904222
Female    0.094721
Other     0.001056
Name: Gender, dtype: float64

In [30]:
{column_name: (df_categorical_columns[column_name].value_counts()/len(df_categorical_columns[column_name])).to_dict() for column_name in df_categorical_columns.columns}

{'Gender': {'Male': 0.9042224582949067,
  'Female': 0.09472105158706996,
  'Other': 0.0010564901180233712},
 'Age range': {'18-24': 0.35860543304420917,
  '25-34': 0.2521544226660907,
  'over 34': 0.21585669090951543,
  '10-17': 0.17287096811851368,
  'under 10': 0.0005124852616710055},
 'Self-defined ethnicity': {'White - English/Welsh/Scottish/Northern Irish/British': 0.4122004179264776,
  'Other ethnic group - Not stated': 0.15952417085955103,
  'White - Any other White background': 0.08685983119505959,
  'Black/African/Caribbean/Black British - Any other Black/African/Caribbean background': 0.07263048528501885,
  'Black/African/Caribbean/Black British - African': 0.051073417307759654,
  'Asian/Asian British - Any other Asian background': 0.04748485308366701,
  'Black/African/Caribbean/Black British - Caribbean': 0.03369561410676971,
  'Asian/Asian British - Pakistani': 0.027904180432168666,
  'Asian/Asian British - Bangladeshi': 0.019621531385344556,
  'Other ethnic group - Any oth

Based on the provided column ratios, we can draw the following conclusions:

1. Gender: The majority of individuals stopped and searched are male (90.42%). Female and other gender individuals account for 9.47% and 0.11%, respectively. This indicates a possible gender bias in the stop and search incidents.

2. Age range: The age group with the highest proportion of stop and search incidents is 18-24 (35.86%), followed by 25-34 (25.22%), over 34 (21.59%), and 10-17 (17.29%). Individuals under 10 years old have the lowest proportion (0.05%).

3. Self-defined ethnicity: The majority of individuals identify as "White - English/Welsh/Scottish/Northern Irish/British" (41.22%). The remaining ethnic groups have significantly lower proportions, with the second largest group being "Other ethnic group - Not stated" (15.95%).

4. Officer-defined ethnicity: According to officers, 59.26% of individuals stopped and searched are White, 23.01% are Black, 13.98% are Asian, 3.39% are from Other ethnic groups, and 0.35% are Mixed.

5. Legislation: The majority of stop and search incidents are conducted under the "Misuse of Drugs Act 1971 (section 23)" (63.85%), followed by the "Police and Criminal Evidence Act 1984 (section 1)" (29.75%).

6. Object of search: The most common object of search is "Controlled drugs" (65.15%), followed by "Offensive weapons" (13.15%) and "Stolen goods" (8.84%).

7. Outcome: In most cases (73.25%), there is no further action taken after the stop and search. Arrests account for 13.05% of the outcomes, while community resolutions account for 7.97%.

8. Outcome linked to object of search: The outcome is linked to the object of search in 13.49% of cases, whereas it is not linked in 16.24% of cases.

9. Removal of more than just outer clothing: In 38.23% of incidents, only outer clothing is removed. In 1.47% of incidents, more than outer clothing is removed.

10. Station: The Metropolitan Police force has the highest proportion of stop and search incidents (51.00%). The remaining police forces have significantly lower proportions, with Merseyside having the second-highest at 4.50%.

These conclusions help to provide a better understanding of the stop and search incidents' demographic distribution and the reasons and outcomes associated with them. This information can be useful when analyzing potential biases and assessing the effectiveness of current policing strategies.

Based on the analysis and conclusions drawn from the data, the following recommendations can be made to address potential biases and improve the effectiveness of stop and search practices:

1. Address gender bias: Investigate the reasons behind the overwhelming majority of stop and search incidents involving males. Develop training and guidelines for officers to ensure that gender bias is minimized in their decision-making processes.

2. Focus on high-risk age groups: Allocate resources and efforts towards addressing issues within the age groups with higher proportions of stop and search incidents (18-24 and 25-34). Implement community engagement programs and educational initiatives to help prevent criminal activity among these age groups.

3. Enhance cultural sensitivity training: Provide officers with comprehensive cultural sensitivity training to better understand various ethnic communities and ensure unbiased policing.

4. Improve data collection on ethnicity: Encourage individuals to self-identify their ethnicity more accurately during stop and search incidents, and ensure that officers record this information consistently to facilitate better analysis of potential biases.

5. Evaluate effectiveness of stop and search tactics: Regularly review the outcomes of stop and search incidents to assess the effectiveness of current tactics and legislation. This includes analyzing the proportion of incidents resulting in arrests, penalties, or other actions, and identifying areas for improvement.

6. Strengthen community relations: Build trust and rapport with communities by increasing transparency in stop and search practices and engaging in dialogues with community members to address their concerns and better understand their perspectives.

7. Conduct regular audits: Perform regular audits of stop and search data to identify potential biases and disparities, and use this information to inform adjustments to policing strategies and practices.

8. Review officer training and guidelines: Continuously review and update officer training and guidelines on stop and search procedures to ensure they are in line with best practices and legal requirements.

9. Encourage accountability: Foster a culture of accountability within police forces by establishing clear expectations and consequences for biased or inappropriate behavior during stop and search incidents.

10. Monitor geographical disparities: Analyze data on stop and search incidents across different police forces to identify geographical disparities and share best practices among forces to address these disparities effectively.

Implementing these recommendations can help address potential biases, enhance the effectiveness of stop and search practices, and ultimately contribute to improving community relations and public trust in the police.
